File: server/internal/content/topic-planning/contract.go
// Package topicplanning owns manually curated topic cards and append-only
// frozen brief revisions.
package topicplanning

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"time"
)

var (
	ErrInvalid  = errors.New("invalid topic planning input")
	ErrNotFound = errors.New("topic card or brief revision not found")
	ErrStorage  = errors.New("topic planning storage unavailable")
)

// FieldError names the field that was wrong, so a 400 can say which one.
type FieldError struct {
	Field  string `json:"field"`
	Reason string `json:"reason,omitempty"`
}

func (e FieldError) Error() string { return e.Reason + ": " + e.Field }
func (e FieldError) Unwrap() error { return ErrInvalid }

type Status string

const (
	StatusDraft    Status = "draft"
	StatusStarted  Status = "started"
	StatusSaved    Status = "saved"
	StatusDeferred Status = "deferred"
	StatusDropped  Status = "dropped"
)

type Action string

const (
	ActionStart Action = "start"
	ActionSave  Action = "save"
	ActionDefer Action = "defer"
	ActionDrop  Action = "drop"
)

func (a Action) Status() (Status, bool) {
	switch a {
	case ActionStart:
		return StatusStarted, true
	case ActionSave:
		return StatusSaved, true
	case ActionDefer:
		return StatusDeferred, true
	case ActionDrop:
		return StatusDropped, true
	default:
		return "", false
	}
}

// AccountFilterNone selects the cards that are linked to no account.
//
// A list filter needs three answers, not two: every card, the cards of one
// account, and the cards nobody has attached to an account yet. The third one
// has no id to name it, so it gets a word an id can never be - ids are hex.
const AccountFilterNone = "none"

type TopicCard struct {
	TopicCardID               string    `json:"topic_card_id"`
	WorkspaceID               string    `json:"workspace_id"`
	AccountID                 *string   `json:"account_id"`
	AudienceProblemJudgment   string    `json:"audience_problem_judgment"`
	IPFit                     string    `json:"ip_fit"`
	Timing                    string    `json:"timing"`
	ExistingContentRelation   string    `json:"existing_content_relation"`
	EvidenceGapsAndInvestment string    `json:"evidence_gaps_and_investment"`
	Channels                  []string  `json:"channels"`
	// SOP 5.2 item 2: "为什么适合这个 IP，引用哪些素材和过去的经营结论".
	FitSourceIDs              []string  `json:"fit_source_ids"`
	// SOP 5.2 item 5: "证据是否充分，存在什么缺口，需要多少研究或制作投入".
	EvidenceSourceIDs         []string  `json:"evidence_source_ids"`
	RecommendedAction         string    `json:"recommended_action"`
	Status                    Status    `json:"status"`
	DecisionReason            string    `json:"decision_reason"`
	DecisionNote              string    `json:"decision_note"`
	StartedBriefRevisionID    *string   `json:"started_brief_revision_id"`
	CreatedAt                 time.Time `json:"created_at"`
	UpdatedAt                 time.Time `json:"updated_at"`
}

// PatchString distinguishes a missing field from an explicitly empty string.
// Empty text is a legitimate "not known" answer on a topic card, while a
// missing field must leave the stored value untouched.
type PatchString struct {
	Set   bool
	Value string
}

func (p *PatchString) UnmarshalJSON(data []byte) error {
	if bytes.Equal(bytes.TrimSpace(data), []byte("null")) {
		return ErrInvalid
	}
	var value string
	if err := json.Unmarshal(data, &value); err != nil {
		return ErrInvalid
	}
	p.Set = true
	p.Value = value
	return nil
}

// TopicBodyPatch is the only mutable portion of an existing topic card.
// Account ownership, source references, decision state, channels and the
// recommended action all have their own semantics and must not be changed by
// this endpoint.
type TopicBodyPatch struct {
	AudienceProblemJudgment   PatchString `json:"audience_problem_judgment"`
	IPFit                     PatchString `json:"ip_fit"`
	Timing                    PatchString `json:"timing"`
	ExistingContentRelation   PatchString `json:"existing_content_relation"`
	EvidenceGapsAndInvestment PatchString `json:"evidence_gaps_and_investment"`
}

func (p *TopicBodyPatch) UnmarshalJSON(data []byte) error {
	type wire TopicBodyPatch
	var decoded wire
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&decoded); err != nil {
		return ErrInvalid
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		return ErrInvalid
	}
	*p = TopicBodyPatch(decoded)
	return nil
}

func (p TopicBodyPatch) Validate() error {
	if !p.AudienceProblemJudgment.Set && !p.IPFit.Set && !p.Timing.Set &&
		!p.ExistingContentRelation.Set && !p.EvidenceGapsAndInvestment.Set {
		return ErrInvalid
	}
	return nil
}

type BriefRevision struct {
	BriefRevisionID      string    `json:"brief_revision_id"`
	TopicCardID          string    `json:"topic_card_id"`
	WorkspaceID          string    `json:"workspace_id"`
	Revision             int64     `json:"revision"`
	Audience             string    `json:"audience"`
	CoreProblem          string    `json:"core_problem"`
	ClaimAndBoundaries   string    `json:"claim_and_boundaries"`
	Channels             []string  `json:"channels"`
	Format               string    `json:"format"`
	Structure            string    `json:"structure"`
	CitationRequirements string    `json:"citation_requirements"`
	SourceScope          string    `json:"source_scope"`
	Deliverable          string    `json:"deliverable"`
	TimeLimit            string    `json:"time_limit"`
	CostLimit            string    `json:"cost_limit"`
	CreatedAt            time.Time `json:"created_at"`
}

type ActionRequest struct {
	Action Action        `json:"action"`
	Reason string        `json:"reason"`
	Note   string        `json:"note"`
	Brief  BriefRevision `json:"brief"`
}

type ActionResult struct {
	TopicCard TopicCard      `json:"topic_card"`
	Brief     *BriefRevision `json:"brief_revision,omitempty"`
}

File: packages/core/content/topic-planning/contract.ts
import { z } from "zod";
import { parseWithFallback } from "@multica/core/api/schema";

// Status is deliberately open-ended on reads. An older installed client can
// receive a state introduced by a newer server and render the generic label
// instead of replacing the whole card with the fallback.
export const topicStatusSchema = z.string();

export const topicCardSchema = z
  .object({
    topic_card_id: z.string(),
    workspace_id: z.string(),
    account_id: z.string().nullable(),
    audience_problem_judgment: z.string(),
    ip_fit: z.string(),
    timing: z.string(),
    existing_content_relation: z.string(),
    evidence_gaps_and_investment: z.string(),
    channels: z.array(z.string()),
    recommended_action: z.string(),
    fit_source_ids: z.array(z.string()).catch([]),
    evidence_source_ids: z.array(z.string()).catch([]),
    status: topicStatusSchema,
    decision_reason: z.string(),
    decision_note: z.string(),
    started_brief_revision_id: z.string().nullable(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .loose()
  .transform((card) => ({
    topicCardId: card.topic_card_id,
    workspaceId: card.workspace_id,
    accountId: card.account_id,
    audienceProblemJudgment: card.audience_problem_judgment,
    ipFit: card.ip_fit,
    timing: card.timing,
    existingContentRelation: card.existing_content_relation,
    evidenceGapsAndInvestment: card.evidence_gaps_and_investment,
    channels: card.channels,
    recommendedAction: card.recommended_action,
    fitSourceIds: card.fit_source_ids,
    evidenceSourceIds: card.evidence_source_ids,
    status: card.status,
    decisionReason: card.decision_reason,
    decisionNote: card.decision_note,
    startedBriefRevisionId: card.started_brief_revision_id,
    createdAt: card.created_at,
    updatedAt: card.updated_at,
  }));

export const briefRevisionSchema = z
  .object({
    brief_revision_id: z.string(),
    topic_card_id: z.string(),
    workspace_id: z.string(),
    revision: z.number().int().positive(),
    audience: z.string(),
    core_problem: z.string(),
    claim_and_boundaries: z.string(),
    channels: z.array(z.string()),
    format: z.string(),
    structure: z.string(),
    citation_requirements: z.string(),
    source_scope: z.string(),
    deliverable: z.string(),
    time_limit: z.string(),
    cost_limit: z.string(),
    created_at: z.string(),
  })
  .loose()
  .transform((brief) => ({
    briefRevisionId: brief.brief_revision_id,
    topicCardId: brief.topic_card_id,
    workspaceId: brief.workspace_id,
    revision: brief.revision,
    audience: brief.audience,
    coreProblem: brief.core_problem,
    claimAndBoundaries: brief.claim_and_boundaries,
    channels: brief.channels,
    format: brief.format,
    structure: brief.structure,
    citationRequirements: brief.citation_requirements,
    sourceScope: brief.source_scope,
    deliverable: brief.deliverable,
    timeLimit: brief.time_limit,
    costLimit: brief.cost_limit,
    createdAt: brief.created_at,
  }));

export const topicCardListSchema = z
  .object({ topic_cards: z.array(topicCardSchema) })
  .loose()
  .transform((list) => ({ topicCards: list.topic_cards }));

export const briefRevisionListSchema = z
  .object({ brief_revisions: z.array(briefRevisionSchema) })
  .loose()
  .transform((list) => ({ briefRevisions: list.brief_revisions }));

export const topicActionResultSchema = z
  .object({
    topic_card: topicCardSchema,
    brief_revision: briefRevisionSchema.optional(),
  })
  .loose()
  .transform((result) => ({
    topicCard: result.topic_card,
    briefRevision: result.brief_revision,
  }));

export type TopicStatus = z.infer<typeof topicStatusSchema>;
export type TopicCard = z.infer<typeof topicCardSchema>;
export type BriefRevision = z.infer<typeof briefRevisionSchema>;
export type TopicActionResult = z.infer<typeof topicActionResultSchema>;

export interface TopicCardInput {
  accountId?: string | null;
  audienceProblemJudgment: string;
  ipFit: string;
  timing: string;
  existingContentRelation: string;
  evidenceGapsAndInvestment: string;
  channels: string[];
  recommendedAction: string;
  fitSourceIds?: string[];
  evidenceSourceIds?: string[];
}

export interface SetContentTopicSourcesInput {
  fitSourceIds?: string[];
  evidenceSourceIds?: string[];
}

// Omission preserves the stored answer. An empty string is deliberate: it
// clears one answer without changing the other four.
export interface TopicBodyPatchInput {
  audienceProblemJudgment?: string;
  ipFit?: string;
  timing?: string;
  existingContentRelation?: string;
  evidenceGapsAndInvestment?: string;
}

export interface BriefRevisionInput {
  audience: string;
  coreProblem: string;
  claimAndBoundaries: string;
  channels: string[];
  format: string;
  structure: string;
  citationRequirements: string;
  sourceScope: string;
  deliverable: string;
  timeLimit: string;
  costLimit: string;
}

export interface TopicActionInput {
  action: "start" | "save" | "defer" | "drop";
  reason?: string;
  note?: string;
  brief?: BriefRevisionInput;
}

export function topicCardInputToWire(
  input: TopicCardInput,
): Record<string, unknown> {
  const wire: Record<string, unknown> = {
    account_id: input.accountId ?? null,
    audience_problem_judgment: input.audienceProblemJudgment,
    ip_fit: input.ipFit,
    timing: input.timing,
    existing_content_relation: input.existingContentRelation,
    evidence_gaps_and_investment: input.evidenceGapsAndInvestment,
    channels: input.channels,
    recommended_action: input.recommendedAction,
  };
  if (input.fitSourceIds !== undefined) {
    wire.fit_source_ids = input.fitSourceIds;
  }
  if (input.evidenceSourceIds !== undefined) {
    wire.evidence_source_ids = input.evidenceSourceIds;
  }
  return wire;
}

export function setContentTopicSourcesInputToWire(
  input: SetContentTopicSourcesInput,
): Record<string, unknown> {
  const wire: Record<string, unknown> = {};
  if (input.fitSourceIds !== undefined) {
    wire.fit_source_ids = input.fitSourceIds;
  }
  if (input.evidenceSourceIds !== undefined) {
    wire.evidence_source_ids = input.evidenceSourceIds;
  }
  return wire;
}

export function topicBodyPatchInputToWire(
  input: TopicBodyPatchInput,
): Record<string, unknown> {
  const wire: Record<string, unknown> = {};
  if (input.audienceProblemJudgment !== undefined) {
    wire.audience_problem_judgment = input.audienceProblemJudgment;
  }
  if (input.ipFit !== undefined) {
    wire.ip_fit = input.ipFit;
  }
  if (input.timing !== undefined) {
    wire.timing = input.timing;
  }
  if (input.existingContentRelation !== undefined) {
    wire.existing_content_relation = input.existingContentRelation;
  }
  if (input.evidenceGapsAndInvestment !== undefined) {
    wire.evidence_gaps_and_investment = input.evidenceGapsAndInvestment;
  }
  return wire;
}

export function briefRevisionInputToWire(
  input: BriefRevisionInput,
): Record<string, unknown> {
  return {
    audience: input.audience,
    core_problem: input.coreProblem,
    claim_and_boundaries: input.claimAndBoundaries,
    channels: input.channels,
    format: input.format,
    structure: input.structure,
    citation_requirements: input.citationRequirements,
    source_scope: input.sourceScope,
    deliverable: input.deliverable,
    time_limit: input.timeLimit,
    cost_limit: input.costLimit,
  };
}

export function topicActionInputToWire(
  input: TopicActionInput,
): Record<string, unknown> {
  return {
    action: input.action,
    reason: input.reason ?? "",
    note: input.note ?? "",
    brief: input.brief ? briefRevisionInputToWire(input.brief) : {},
  };
}

export const EMPTY_TOPIC_CARD: TopicCard = {
  topicCardId: "",
  workspaceId: "",
  accountId: null,
  audienceProblemJudgment: "",
  ipFit: "",
  timing: "",
  existingContentRelation: "",
  evidenceGapsAndInvestment: "",
  channels: [],
  recommendedAction: "",
  fitSourceIds: [],
  evidenceSourceIds: [],
  status: "draft",
  decisionReason: "",
  decisionNote: "",
  startedBriefRevisionId: null,
  createdAt: "",
  updatedAt: "",
};

export const EMPTY_BRIEF_REVISION: BriefRevision = {
  briefRevisionId: "",
  topicCardId: "",
  workspaceId: "",
  revision: 1,
  audience: "",
  coreProblem: "",
  claimAndBoundaries: "",
  channels: [],
  format: "",
  structure: "",
  citationRequirements: "",
  sourceScope: "",
  deliverable: "",
  timeLimit: "",
  costLimit: "",
  createdAt: "",
};

export function parseTopicCard(data: unknown): TopicCard {
  return parseWithFallback(data, topicCardSchema, EMPTY_TOPIC_CARD, {
    endpoint: "content-topics/detail",
  });
}

export function parseTopicCards(data: unknown): TopicCard[] {
  return parseWithFallback(
    data,
    topicCardListSchema,
    { topicCards: [] as TopicCard[] },
    { endpoint: "content-topics/list" },
  ).topicCards;
}

export function parseBriefRevision(data: unknown): BriefRevision {
  return parseWithFallback(data, briefRevisionSchema, EMPTY_BRIEF_REVISION, {
    endpoint: "content-topics/brief",
  });
}

export function parseBriefRevisions(data: unknown): BriefRevision[] {
  return parseWithFallback(
    data,
    briefRevisionListSchema,
    { briefRevisions: [] as BriefRevision[] },
    { endpoint: "content-topics/briefs" },
  ).briefRevisions;
}

export function parseTopicActionResult(data: unknown): TopicActionResult {
  return parseWithFallback<TopicActionResult>(
    data,
    topicActionResultSchema,
    { topicCard: EMPTY_TOPIC_CARD, briefRevision: undefined },
    { endpoint: "content-topics/action" },
  );
}

export function topicStatusLabel(status: string): string {
  switch (status) {
    case "draft":
      return "Draft";
    case "started":
      return "Started";
    case "saved":
      return "Saved";
    case "deferred":
      return "Deferred";
    case "dropped":
      return "Dropped";
    default:
      return "Unknown";
  }
}
